File: internal/handler/list.go
package handler

import (
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
	"github.com/webgress/s3gateway/internal/s3response"
	"github.com/webgress/s3gateway/internal/storage"
)

type ListHandler struct {
	fs *storage.Filesystem
}

func NewListHandler(fs *storage.Filesystem) *ListHandler {
	return &ListHandler{fs: fs}
}

func (h *ListHandler) ListObjectsV2(w http.ResponseWriter, r *http.Request) {
	bucket := mux.Vars(r)["bucket"]

	// Check bucket exists
	if err := h.fs.HeadBucket(bucket); err != nil {
		if err == storage.ErrBucketNotFound {
			s3response.WriteError(w, r, s3response.ErrNoSuchBucket)
			return
		}
		s3response.WriteError(w, r, s3response.ErrInternalError)
		return
	}

	prefix := r.URL.Query().Get("prefix")
	delimiter := r.URL.Query().Get("delimiter")
	startAfter := r.URL.Query().Get("start-after")
	continuationToken := r.URL.Query().Get("continuation-token")

	maxKeys := 1000
	if mk := r.URL.Query().Get("max-keys"); mk != "" {
		v, err := strconv.Atoi(mk)
		if err == nil && v >= 0 {
			maxKeys = v
		}
	}

	input := storage.ListObjectsInput{
		Bucket:            bucket,
		Prefix:            prefix,
		Delimiter:         delimiter,
		MaxKeys:           maxKeys,
		StartAfter:        startAfter,
		ContinuationToken: continuationToken,
	}

	output, err := h.fs.ListObjects(input)
	if err != nil {
		s3response.WriteError(w, r, s3response.ErrInternalError)
		return
	}

	result := s3response.ListBucketResultV2{
		Name:                  bucket,
		Prefix:                prefix,
		Delimiter:             delimiter,
		MaxKeys:               maxKeys,
		IsTruncated:           output.IsTruncated,
		KeyCount:              len(output.Objects) + len(output.CommonPrefixes),
		StartAfter:            startAfter,
		ContinuationToken:     continuationToken,
		NextContinuationToken: output.NextContinuationToken,
	}

	for _, obj := range output.Objects {
		result.Contents = append(result.Contents, s3response.ObjectEntry{
			Key:          obj.Key,
			LastModified: s3response.FormatTime(obj.LastModified),
			ETag:         obj.ETag,
			Size:         obj.Size,
			StorageClass: "STANDARD",
		})
	}

	for _, p := range output.CommonPrefixes {
		result.CommonPrefixes = append(result.CommonPrefixes, s3response.CommonPrefix{
			Prefix: p,
		})
	}

	s3response.WriteXML(w, http.StatusOK, result)
}

File: rust/src/handler/list.rs
//! ListObjectsV2 handler: GET /{bucket}?list-type=2 (and plain GET /{bucket}).

use hyper::Response;

use crate::s3response::{format_time, ListBucketResultV2, ObjectEntry, S3ErrorCode};
use crate::storage::ListObjectsInput;

use super::{map_storage_error, xml_ok, Ctx, HandlerRequest, RespBody};

/// GET /{bucket} (with or without list-type=2) — ListObjectsV2.
pub async fn list_objects_v2(
    ctx: &Ctx,
    req: HandlerRequest,
) -> Result<Response<RespBody>, S3ErrorCode> {
    let prefix = req.query1("prefix").unwrap_or("").to_string();
    let delimiter = req.query1("delimiter").unwrap_or("").to_string();
    let start_after = req.query1("start-after").unwrap_or("").to_string();
    let continuation_token = req.query1("continuation-token").unwrap_or("").to_string();
    // F14: preserve ABSENT vs explicit value. `None` -> storage defaults to 1000;
    // `Some(0)` -> empty page with IsTruncated. The wire `max-keys` echoed back in
    // the response uses the effective value (1000 when absent).
    let max_keys: Option<i32> = req.query1("max-keys").and_then(|s| s.parse().ok());
    let effective_max_keys = max_keys.unwrap_or(1000);

    let input = ListObjectsInput {
        bucket: req.bucket.clone(),
        prefix: prefix.clone(),
        delimiter: delimiter.clone(),
        max_keys,
        start_after: start_after.clone(),
        continuation_token: continuation_token.clone(),
    };

    let fs = ctx.fs.clone();
    let out = tokio::task::spawn_blocking(move || fs.list_objects(&input))
        .await
        .map_err(|_| S3ErrorCode::InternalError)?
        .map_err(|e| map_storage_error(&e))?;

    let key_count = (out.objects.len() + out.common_prefixes.len()) as i32;
    let result = ListBucketResultV2 {
        name: req.bucket.clone(),
        prefix,
        delimiter,
        max_keys: effective_max_keys,
        is_truncated: out.is_truncated,
        key_count,
        start_after,
        continuation_token,
        next_continuation_token: out.next_continuation_token,
        contents: out
            .objects
            .into_iter()
            .map(|o| ObjectEntry {
                key: o.key,
                last_modified: format_time(o.last_modified_unix),
                etag: o.etag,
                size: o.size,
                storage_class: "STANDARD".to_string(),
            })
            .collect(),
        common_prefixes: out.common_prefixes,
    };

    Ok(xml_ok(result.to_xml()))
}
